### PR TITLE
SI-8835 Fix implementation of Iterator drop to remove quadratic behavior

### PR DESCRIPTION
**Pull request by erikerlandson**
The Iterator drop method (defined in Iterator.scala) generates nested anonymous classes. When drop is used in a loop, this causes linear traversal to become quadratic, and will also eventually cause a stack overflow if invoked sufficiently many times.

I wrote up a description and analysis of the problem here:
http://erikerlandson.github.io/blog/2014/09/03/matryoshka-class-construction-from-the-scala-iterator-drop-method/

### CODE REVIEW

**Comment by som-snytt** [comment]
This fixes the problem only for drop and not for the other anon AbstractIterators that reference self.

Here is a sample fix that eliminates the outer ref:

```
import collection.AbstractIterator

trait SlicedIterator[+A] extends Iterator[A] {
  protected def self: Iterator[A] = this

  override def slice(from: Int, until: Int): Iterator[A] = {
    val lo = from max 0
    var toDrop = lo
    while (toDrop > 0 && self.hasNext) {
      self.next()
      toDrop -= 1
    }

    class AI[+A](override val self: Iterator[A]) extends AbstractIterator[A] with SlicedIterator[A] {
      private var remaining = until - lo
      def hasNext = remaining > 0 && self.hasNext
      def next(): A =
        if (remaining > 0) {
          remaining -= 1
          self.next()
        }
        else Iterator.empty.next()
    }
    new AI(self)
  }
}
```

which seems to work for your linked test with

```
class TestIterator[T: ClassTag](val iter: Iterator[T]) extends Iterator[T] with SlicedIterator[T]
```

That's just a sample. A better name for `self` would be `underlying`.

And I haven't tested it; I'm just on lunch break.

Is it too bad that the dropping part is eager?  Just curious.

**Comment by erikerlandson** [comment]
hi @som-snytt,

I'm not quite following how your example addresses the Iterator.scala implementations.   

My PR doesn't address slice or take, although I don't see how those can get away _without_ defining a new class, as it has to contain the new stopping point for future reference, unlike drop.

Making it somehow lazy instead of eager would be nice, although I can't see how in the general case, which is what Iterator.scala covers.   I don't see it as a big problem, as it is only eager with elements that we are by definition discarding.  Remaining elements are left uncomputed.

**Comment by som-snytt** [comment]
Slice has to create a class, but that class `AI` only has to hold onto the original, underlying iterator, not the parent AI that created it.  My idea was that you don't have a call chain parent.next and also that you lose the outer pointer which would keep the anon classes from getting GC'd.

**Comment by erikerlandson** [comment]
@som-snytt, I tried a couple different variations on your example above, and they both retain the quadratic behavior when I run the `drop_time()` function in my blog post.

I recorded the variations I tried here:
https://gist.github.com/erikerlandson/eb7edf947771abb33ff6

However, I'm still not feeling 100% sure I understand the theory behind how your example is _intended_ to work, so maybe I didn't adapt it right.

**Comment by som-snytt** [comment]
I wasn't set up at lunch to try a build, but here's a gist.  It has the patch to Iterator and your test rig, modified.

https://gist.github.com/som-snytt/e4c47afb5c8f2b943f77

I remembered to reuse the stack trace truncator from the REPL.  And I also found a use for an idea the other day to use f"$x%#s" to mean hashCode, which is easy to fake; I'd also wanted f"$x%##" to distinguish identity hash.

It turns out adding a member `underlying` to Iterator is bad because everyone who writes an anonymous iterator has a notion of underlying, often called `underlying`, so Wrappers broke and also Global, ha.  Here I just called it `delegate`.

I have no idea if this works for similar idioms in Iterator or if it makes something else blow up, as is so easy to do when messing around.

**Comment by som-snytt** [comment]
I think a junit test like the following, or similar:

https://github.com/som-snytt/scala/blob/428528fa4a1201891fb5fa70b5a73fceb21d6ad9/test/junit/scala/collection/IteratorTest.scala#L29

**Comment by erikerlandson** [comment]
@som-snytt, thanks for your feedback!   I modified slice as you described, and added unit testing.  It performs in what looks like linear time with some spikes I suspect are GC related.  I did end up keeping my original definition for drop, because it performs better than when it calls slice.

Example of the kind of performance I'm seeing with the new slice and drop:

scala> for (j <- 1 to 100) yield slice_time((1 to j \* 5000).toList.iterator)
res3: scala.collection.immutable.IndexedSeq[Long] = Vector(0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 13, 1, 1, 1, 1, 19, 1, 1, 2, 25, 1, 1, 1, 1, 2, 1, 2, 2, 1, 2, 1, 34, 2, 2, 2, 2, 2, 2, 3, 3, 48, 2, 54, 3, 3, 3, 3, 2, 3, 2, 3, 3, 3, 3, 3, 3, 4, 3, 3, 3, 4, 3, 3, 3, 4, 4, 73, 3, 86, 4, 5, 4, 4, 4, 4, 79, 5, 4, 75, 5, 4, 63, 5, 5, 5, 5, 5, 5, 72, 90, 85, 72, 69, 81, 100)

scala> for (j <- 1 to 100) yield drop_time((1 to j \* 5000).toList.iterator)
res4: scala.collection.immutable.IndexedSeq[Long] = Vector(0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 2, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 1, 2, 1, 7, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2)

**Comment by erikerlandson** [comment]
Tangentially, a useful side effect of this patch is that `drop` is no longer secretly capped by Int.MaxValue, which seems right because `drop` implies no upper bound on its result.

**Comment by adriaanm** [comment]
I apologize we haven't been better at following up here! @Ichoran could you take a look?

**Comment by adriaanm** [comment]
In the mean time, Welcome, @erikerlandson! Could you please sign the [Scala CLA](http://typesafe.com/contribute/cla/scala)?

**Comment by gkossakowski** [comment]
This PR doesn't merge cleanly anymore so it will require a rebase. I'm scheduling it for 2.11.4 so we have enough of time to get the kinks ironed out.

**Comment by erikerlandson** [comment]
@adriaanm & @gkossakowski, sorry I was on vacation last week.   I rebased the PR and signed the CLA.

**Comment by erikerlandson** [comment]
The failed builds seem to have broken in some previous build, if I'm reading the output right:
Status: broken since build #4047@946dd5>2.11.x [8725]
@gkossakowski

**Comment by gourlaysama** [comment]
@erikerlandson the failures are spurious, see https://github.com/scala/scala/pull/4046#issuecomment-58663397 (although it doesn't appear to be fixed...) /cc @dragos.

File: src/library/scala/collection/Iterator.scala [hunk 1]
```diff
@@ -320,7 +320,16 @@ trait Iterator[+A] extends TraversableOnce[A] {
...
+  def drop(n: Int): Iterator[A] = {
+    var j = 0
+    while (j < n && this.hasNext) {
+      this.next
+      j += 1
+    }
+    this
+  }
+
+  protected def delegate: Iterator[A] = this
```
Comment (Ichoran) [inline src/library/scala/collection/Iterator.scala]: This isn't going to be binary compatible.  You're inviting people to override it with protected, but if they do and then use 2.11.2...BOOM!  private[collection] is better...someone else will have to comment whether this is good enough.  (Removed part of comment--answer was: it's a trait, silly!)  I'm also not entirely clear why you need this method at all.  Why not have SliceIterator define it and pattern match it out if it's a SliceIterator of a SliceIterator?

**Comment by erikerlandson** [comment]
I experimented with a variation on `slice` that did not use the 'delegate' technique to address @Ichoran 's concerns.   However, while adding some further unit tests I discovered that both "nesting free" variations of `slice` fail this test:

`Iterator.from(0).slice(3, 7).slice(1, 10).toSeq sameElements Seq(4, 5, 6)`

In particular, it fails because the "nest-free" logic works by referring to the original iterator, and so the slice ending condition in `hasNext` for `slice(3, 7)` gets ignored.

I restored the original implementation of `slice` (which also gets rid of the need for all the whitelist/blacklist hackery).  I also beefed up the unit testing for both `drop` and `slice`

**Comment by Ichoran** [comment]
You seem to have an extraneous import of `Thread`.

**Comment by som-snytt** [comment]
Parking this point of reference here:

https://github.com/scala/scala/commit/783721e98a4f61fc0c72298811b15e99c058d5e6

**Comment by Ichoran** [comment]
LGTM

**Comment by som-snytt** [comment]
This is nit-picky, but everywhere `next()` is used with the "danger, danger, side-effects" parens.

...everywhere except the late-coming Concat- and Join- subclasses.

**Comment by Ichoran** [comment]
@som-snytt makes the important point (elsewhere!) that this changes the laziness of drop.  This isn't an acceptable change for a point release, and given that Iterator is often used for its laziness, I don't think it's a good idea at all.  So I think we're going to have to nix this change for 2.11.  We can discuss whether this is the right thing to do for 2.12, but I think not.

As a workaround, if you really just need to eagerly discard elements, create a `discard` extension method:

```
implicit class EnableIteratorDiscarding[A](private val underlying: Iterator[A]) extends AnyVal {
  def discard(n: Int) = {
    var k = n
    while (k > 0 && underlying.hasNext) { k -= 1; underlying.next }
    underlying
  }
}
```

**Comment by som-snytt** [comment]
Not sure that it changes laziness, but you make the important point that
changing to make it lazy would also be no-go for 2.11.

Elsewhere, I moved the iterator tests to the junit test, which I'll PR.
Maybe I'll then PR the lazy SliceIterator to 2.12.  It just tracks the drop
count and limit, and doesn't suffer from the bug pointed out earlier, and
is a cleaner impl.

**Comment by Ichoran** [comment]
In particular, this test should pass:

```
val i = Iterator(1,2,3)
val j = i drop 2
i sameElements Iterator(1,2,3)
```

**Comment by erikerlandson** [comment]
@som-snytt & @Ichoran,  I am a bit confused by this issue of laziness.   AFAICT, my implementation of `drop` is neither more or less lazy than what's in there now.  `data.drop(n)` drops the first (n) elements ("eagerly") in both cases.  In the existing implementation, that happens inside the call to `slice`.

**Comment by Ichoran** [comment]
It's more complicated than that due to overriding.  Goodness.  This is
rather a mess.

scala> val i = List(1,2,3).iterator
i: Iterator[Int] = non-empty iterator

scala> val j = i slice(1,2)
j: Iterator[Int] = non-empty iterator

scala> i.foreach(println)
2
3

scala> val i = Array(1,2,3).iterator
i: Iterator[Int] = non-empty iterator

scala> val j = i slice(1,2)
j: Iterator[Int] = non-empty iterator

scala> i.foreach(println)
1
2
3

**Comment by Ichoran** [comment]
LGTM on third look, as it doesn't change behavior.  But we really need to revisit this for 2.12.  I'll add collections tests that highlight that the behavior is inconsistent.

**Comment by erikerlandson** [comment]
@Ichoran, my understanding of iterator semantics was that if you call `drop` (or `slice`, and some others) on an iterator, that it was considered to "invalidate" the original:

"It is of particular importance to note that, unless stated otherwise, one should never use an iterator after calling a method on it. "
http://www.scala-lang.org/api/current/index.html#scala.collection.Iterator

Tangentially, I assumed that was the meaning of: `$consumesAndProducesIterator`, but maybe I got that wrong.

**Comment by som-snytt** [comment]
Ditto to @erikerlandson so LGTMT.  Oh, modulo `next()` with parens.  Maybe we'll get that next time.

**Comment by som-snytt** [comment]
I noticed that my test refactor includes a test that the stack doesn't grow under chained drops. Which was kind of the point.

**Comment by adriaanm** [comment]
> Maybe we'll get that next time.

Did you mean `time.next()`?

**Comment by erikerlandson** [comment]
@som-snytt, I never did do the stack test for the revised `drop`, mostly because its implementation seemed obvious enough to not require it.  However, I could add that, as a guard against anybody in the future thinking "Hey, I can just replace this with a call to `slice` ..."   :)

**Comment by som-snytt** [comment]
@erikerlandson hopefully my PR for the tests will get approved.  You may smile emoticonically, but there has been recidivism around these issues of extension.

**Comment by retronym** [comment]
I've lost track of the status of this PR. Is it ready for merge?

**Comment by som-snytt** [comment]
@Ichoran 's third-look LGTM is still G.

**Comment by Ichoran** [comment]
LGTM still

**Pull request merged**